spec/unit/functions/lookup_spec.rb

```diff
@@ -1,733 +1,1253 @@
-              :type => :merge
+                {
+                  :value => { 'k1' => 'env_e1', 'k2' => 'module_e2', 'k3' => 'env_e3' },
+                  :event => :result,
+                  :merge => :deep,
+                  :options => { 'merge_hash_arrays' => true },
+                  :type => :merge,
+                  :branches => [
+                    {
+                      :key => 'abc::e',
+                      :event => :not_found,
+                      :type => :global,
+                      :name => :hiera
+                    },
+                    {
+                      :type => :data_provider,
+                      :name => 'Environment Data Provider (lookup version 5)',
+                      :branches => [
+                        {
+                          :type => :data_provider,
+                          :name => 'Legacy function "environment::data"',
+                          :key => 'abc::e',
+                          :value => { 'k1' => 'env_e1', 'k3' => 'env_e3' },
+                          :event => :found
+                        }
+                      ]
+                    },
+                    {
+                      :type => :data_provider,
+                      :name => 'Module "abc" Data Provider (lookup version 5)',
+                      :module => 'abc',
+                      :branches => [
+                        {
+                          :type => :data_provider,
+                          :name => 'Legacy function "abc::data"',
+                          :key => 'abc::e',
+                          :event => :found,
+                          :value => {
+                            'k1' => 'module_e1',
+                            'k2' => 'module_e2'
+                          }
+                        }
+                      ]
+                    }
+                  ]
+                }
+              ]
             }
           )
+        end
       end
-    end
 
-    it 'will explain that "lookup_options" is an invalid key' do
-      assemble_and_compile('${r}', "'abc::a'") do |scope|
-        lookup_invocation = Puppet::Pops::Lookup::Invocation.new(scope, {}, {}, true)
-        begin
-          Puppet::Pops::Lookup.lookup('lookup_options', nil, nil, false, nil, lookup_invocation)
-        rescue Puppet::Error
-        end
-        expect(lookup_invocation.explainer.explain).to eq(<<EOS)
+      it 'will explain that "lookup_options" is an invalid key' do
+        assemble_and_compile('${r}', "'abc::a'") do |scope|
+          lookup_invocation = Puppet::Pops::Lookup::Invocation.new(scope, {}, {}, true)
+          begin
+            Puppet::Pops::Lookup.lookup('lookup_options', nil, nil, false, nil, lookup_invocation)
+          rescue Puppet::Error
+          end
+          expect(lookup_invocation.explainer.explain).to eq(<<EOS)
 Invalid key "lookup_options"
 EOS
+        end
       end
-    end
 
-    it 'will explain that "lookup_options" is
```

Comment:
The tests are written from a test implementation standpoint rather than feature/example standpoint. That makes these tests hard to read and map to the complete set of features / desired behavior.

* "Data in the environment is found"
   * "When there is a lookup.conf"
     * "In which case a "hiera.yaml" is ignored
  * "When there is a hiera.yaml (but no lookup.conf)"
* "Old constructs are deprecated such that"
  * "Use of a hiera.yaml v4 issues deprecation"
  * "Use of function data provider issues deprecation with helpful message what to change to"
  * "Use of ... data provider..."
* "Data in a module is found"
  * "..."

etc